### PR TITLE
[plugin-web-app-to-rest-api] Support scheme shortcuts in URL

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java [hunk 1]
```diff
@@ -216,6 +218,14 @@ private String getCssSelector(Element element)
     private URI resolveUri(String uri) throws URISyntaxException
     {
         URI uriToCheck = new URI(uri);
+
+        if (uri.startsWith("//") && uriToCheck.getHost() != null)
+        {
+            String scheme = Optional.ofNullable(webApplicationConfiguration.getMainApplicationPageUrlUnsafely())
+                    .map(URI::getScheme).orElse(HTTPS);
```
Comment (valfirst) [inline vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:225]: why is default value `https`?

File: vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java [hunk 2]
```diff
@@ -216,6 +218,14 @@ private String getCssSelector(Element element)
     private URI resolveUri(String uri) throws URISyntaxException
     {
         URI uriToCheck = new URI(uri);
+
+        if (uri.startsWith("//") && uriToCheck.getHost() != null)
```
Comment (valfirst) [inline vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:222]: when does URL without host start from `//`?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java (hunk 1 above)
Comment (uarlouski) [reply vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:225]: what would you suggest if not https? this is the most widely used

File: vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java (hunk 2 above)
Comment (uarlouski) [reply vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:222]: I think someone might have relative url starting with //, just an assumption

File: vividus-plugin-web-app-to-rest-api/src/test/java/org/vividus/steps/integration/ResourceCheckStepsTests.java [hunk 3]
```diff
@@ -485,10 +499,12 @@ void shouldFilterResourceByRegExpCheckDesiredResourcesAnPostAttachment()
 
     private void runExecutor() throws InterruptedException, ExecutionException
     {
-        doNothing().when(executor).execute(argThat(r -> {
```
Comment (uarlouski) [inline vividus-plugin-web-app-to-rest-api/src/test/java/org/vividus/steps/integration/ResourceCheckStepsTests.java:488]: junit fails with unnecessary stubbing if we throw an exception inside executor, so changed to doAnswer

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java [hunk 4]
```diff
@@ -216,9 +217,19 @@ private String getCssSelector(Element element)
...
+
         if (isNotAbsolute(uriToCheck))
         {
             URI mainApplicationPageUrl = webApplicationConfiguration.getMainApplicationPageUrlUnsafely();
+            if (uri.startsWith("//"))
+            {
+                String scheme = Optional.ofNullable(mainApplicationPageUrl)
+                                        .map(URI::getScheme)
+                                        .orElseThrow(() -> new IllegalStateException(
+                                                String.format("Unable to evaluate protocol for %s url", uri)));
```
[the remark this answers is not in the document]
Comment (uarlouski) [reply vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:229]: I just thought its a bit weird as per morning discussion :)

File: vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java [hunk 5]
```diff
@@ -216,9 +217,20 @@ private String getCssSelector(Element element)
     private URI resolveUri(String uri) throws URISyntaxException
     {
         URI uriToCheck = new URI(uri);
+
         if (isNotAbsolute(uriToCheck))
         {
             URI mainApplicationPageUrl = webApplicationConfiguration.getMainApplicationPageUrlUnsafely();
+            if (uri.charAt(1) == '/')
```
Comment (valfirst) [inline vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:224]: what if uri is `/`

File: vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java [hunk 6]
```diff
@@ -216,9 +217,20 @@ private String getCssSelector(Element element)
...
+            {
+                String scheme = Optional.ofNullable(mainApplicationPageUrl)
+                                        .map(URI::getScheme)
+                                        .orElseThrow(() -> new IllegalStateException(
+                                                "Unable to evaluate protocol for main application page url: "
+                                                        + mainApplicationPageUrl));
+                return new URIBuilder(uri).setScheme(scheme).build();
+            }
+
             if (mainApplicationPageUrl != null)
```
Comment (valfirst) [inline vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:234]: this check is duplicated: here and `Optional.ofNullable`

Comment (uarlouski) [reply vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:234]: but we will never get to this point if we got to that check so I think its not critical

File: vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java [hunk 7]
```diff
@@ -216,9 +217,20 @@ private String getCssSelector(Element element)
...
             {
                 return buildNewUrl(mainApplicationPageUrl, uri);
```
Comment (valfirst) [inline vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:236]:
```java
    private URI resolveUri(String uri) throws URISyntaxException
    {
        URI uriToCheck = new URI(uri);

        if (isNotAbsolute(uriToCheck))
        {
            URI mainApplicationPageUrl = webApplicationConfiguration.getMainApplicationPageUrlUnsafely();

            if (mainApplicationPageUrl != null)
            {
                if (uri.length() <= 1 || uri.charAt(1) != '/')
                {
                    return buildNewUrl(mainApplicationPageUrl, uri);
                }
                String scheme = mainApplicationPageUrl.getScheme();
                if (scheme != null)
                {
                    return new URIBuilder(uri).setScheme(scheme).build();
                }
            }
        }
        return uriToCheck;
    }
```
?

Comment (uarlouski) [reply vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:236]: But it misses the point, I want to throw an exception if we got // URL and we can't determine the scheme, in your code the request to // will be still sent

Comment (uarlouski) [reply vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/steps/integration/ResourceCheckSteps.java:236]: I'll take it but it needs to be reworked on exception in the future

**Review by valfirst — APPROVED** [review]

**Review by valfirst — APPROVED** [review]

**Pull request merged**